File: examples/performance-analysis/simple-agent/weather-profile.go
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"os"
	"runtime/pprof"
	"runtime/trace"
	"time"

	_ "github.com/agenticgokit/agenticgokit/plugins/llm/ollama"
	vnext "github.com/agenticgokit/agenticgokit/v1beta"
)

var (
	cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
	memprofile = flag.String("memprofile", "", "write memory profile to file")
	tracefile  = flag.String("trace", "", "write runtime trace to file")
	verbose    = flag.Bool("v", false, "verbose timing output")
	city       = flag.String("city", "", "city to query (if empty, tests multiple cities)")
	reasoning  = flag.Bool("reasoning", false, "enable agent reasoning loops (default: false for fast path)")
)

// WeatherTool is a minimal LangChain-style tool.
type WeatherTool struct{}

func (t *WeatherTool) Name() string { return "check_weather" }
func (t *WeatherTool) Description() string {
	return "Return the weather forecast for the specified location"
}

// JSONSchema provides the schema for native tool calling.
func (t *WeatherTool) JSONSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"location": map[string]interface{}{
				"type":        "string",
				"description": "City name or common abbreviation (e.g., 'sf'→'San Francisco', 'nyc'→'New York', 'la'→'Los Angeles')",
				"examples":    []string{"San Francisco", "sf", "New York", "nyc", "Tokyo"},
			},
		},
		"required": []string{"location"},
	}
}

func (t *WeatherTool) Execute(ctx context.Context, args map[string]interface{}) (*vnext.ToolResult, error) {
	location, ok := args["location"].(string)
	if !ok || location == "" {
		return &vnext.ToolResult{Success: false, Error: "location parameter required"}, fmt.Errorf("location required")
	}

	forecast := fmt.Sprintf("It's always sunny in %s ☀️", location)
	return &vnext.ToolResult{
		Success: true,
		Content: map[string]interface{}{
			"location": location,
			"forecast": forecast,
		},
	}, nil
}

// Register the tool (automatic discovery).
func init() {
	vnext.RegisterInternalTool("check_weather", func() vnext.Tool { return &WeatherTool{} })
}

// No manual extraction or direct-call fallback; rely on native tool calling.

func main() {
	flag.Parse()

	// Start CPU profiling if requested
	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal("could not create CPU profile: ", err)
		}
		defer f.Close()
		if err := pprof.StartCPUProfile(f); err != nil {
			log.Fatal("could not start CPU profile: ", err)
		}
		defer pprof.StopCPUProfile()
	}

	// Start runtime trace if requested (helps show two LLM calls in reasoning mode)
	if *tracefile != "" {
		f, err := os.Create(*tracefile)
		if err != nil {
			log.Fatal("could not create trace file: ", err)
		}
		defer f.Close()
		if err := trace.Start(f); err != nil {
			log.Fatal("could not start trace: ", err)
		}
		defer trace.Stop()
	}

	startTime := time.Now()
	if *verbose {
		fmt.Printf("[TIMING] Program start\n")
	}

	ctx := context.Background()

	buildStart := time.Now()
	agent, err := vnext.NewBuilder("weather-agent").
		WithConfig(&vnext.Config{
			Name:         "weather-agent",
			SystemPrompt: `You are a helpful assistant.ODnt ask followup questions.`,
			LLM: vnext.LLMConfig{
				Provider:    "ollama",
				Model:       "granite4:latest",
				Temperature: 0.0,
				MaxTokens:   150,
			},
			Tools: &vnext.ToolsConfig{ // Ensure tools are attached to the agent
				Enabled: true,
			},
			// Disable memory for this minimal demo to avoid cross-query bleed
			Memory:  &vnext.MemoryConfig{Enabled: false},
			Timeout: 30 * time.Second,
		}).
		WithPreset(vnext.ChatAgent).                // Tools auto-available
		WithTools(vnext.WithReasoning(*reasoning)). // Apply reasoning toggle
		Build()
	if err != nil {
		log.Fatal(err)
	}
	if *verbose {
		fmt.Printf("[TIMING] Agent build took: %v\n", time.Since(buildStart))
	}

	// Show reasoning mode
	reasoningMode := "disabled (fast path)"
	if *reasoning {
		reasoningMode = "enabled (multi-step reasoning)"
	}
	fmt.Printf("Reasoning mode: %s\n", reasoningMode)

	// Test single city or multiple cities
	var cities []string
	if *city != "" {
		cities = []string{*city}
	} else {
		cities = []string{"sf", "nyc", "tokyo", "london", "paris", "sydney"}
	}
	var totalRunTime time.Duration

	if len(cities) > 1 {
		fmt.Printf("\n=== Testing %d cities ===\n", len(cities))
	}
	for i, cityName := range cities {
		runStart := time.Now()
		query := fmt.Sprintf("will it rain in %s", cityName)
		res, err := agent.Run(ctx, query)
		runDuration := time.Since(runStart)
		totalRunTime += runDuration

		if err != nil {
			fmt.Printf("   [%d] Error for %s: %v\n", i+1, cityName, err)
			continue
		}

		if len(cities) > 1 {
			fmt.Printf("   [%d] %s: %v (took %v)\n", i+1, cityName, res, runDuration)
		} else {
			fmt.Printf("   %s: %v\n", cityName, res)
		}
	}

	fmt.Printf("\n=== Summary ===\n")
	fmt.Printf("   Total calls: %d\n", len(cities))
	fmt.Printf("   Total time: %v\n", totalRunTime)
	fmt.Printf("   Average per call: %v\n", totalRunTime/time.Duration(len(cities)))

	if *verbose {
		fmt.Printf("[TIMING] Total execution: %v\n", time.Since(startTime))
	}

	// Write memory profile if requested
	if *memprofile != "" {
		f, err := os.Create(*memprofile)
		if err != nil {
			log.Fatal("could not create memory profile: ", err)
		}
		defer f.Close()
		if err := pprof.WriteHeapProfile(f); err != nil {
			log.Fatal("could not write memory profile: ", err)
		}
	}
}

File: examples/performance-analysis/langchain-agent/weather-profile.py
import argparse
import cProfile
import pstats
import time
import tracemalloc
from io import StringIO

from langchain.agents import create_agent


def check_weather(location: str) -> str:
    '''Return the weather forecast for the specified location.'''
    return f"It's always sunny in {location} ☀️"


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--cpuprofile', help='write CPU profile to file')
    parser.add_argument('--memprofile', help='write memory profile to file')
    parser.add_argument('-v', '--verbose', action='store_true', help='verbose timing output')
    parser.add_argument('--city', help='city to query (if empty, tests multiple cities)')
    args = parser.parse_args()

    # Start profiling
    profiler = None
    if args.cpuprofile:
        profiler = cProfile.Profile()
        profiler.enable()
    
    if args.memprofile:
        tracemalloc.start()

    start_time = time.time()
    if args.verbose:
        print("[TIMING] Program start")

    # Build agent once
    build_start = time.time()
    graph = create_agent(
        model="ollama:granite4",
        tools=[check_weather],
        system_prompt="You are a helpful assistant. Don't ask follow-up questions.",
    )
    build_duration = time.time() - build_start
    if args.verbose:
        print(f"[TIMING] Agent build took: {build_duration:.3f}s")

    # Test single city or multiple cities
    if args.city:
        cities = [args.city]
    else:
        cities = ["sf", "nyc", "tokyo", "london", "paris", "sydney"]
    total_run_time = 0.0

    if len(cities) > 1:
        print(f"\n=== Testing {len(cities)} cities ===")
    
    for i, city in enumerate(cities, 1):
        run_start = time.time()
        query = f"will it rain in {city}"
        inputs = {"messages": [{"role": "user", "content": query}]}
        
        response = None
        for chunk in graph.stream(inputs, stream_mode="updates"):
            response = chunk
        
        run_duration = time.time() - run_start
        total_run_time += run_duration
        
        if len(cities) > 1:
            print(f"   [{i}] {city}: {response} (took {run_duration:.3f}s)")
        else:
            print(f"   {city}: {response}")

    print(f"\n=== Summary ===")
    print(f"   Total calls: {len(cities)}")
    print(f"   Total time: {total_run_time:.3f}s")
    print(f"   Average per call: {total_run_time/len(cities):.3f}s")

    if args.verbose:
        print(f"[TIMING] Total execution: {time.time() - start_time:.3f}s")

    # Stop and save profiling data
    if profiler:
        profiler.disable()
        profiler.dump_stats(args.cpuprofile)
        print(f"\nCPU profile saved to {args.cpuprofile}")
        
        # Print summary to console
        s = StringIO()
        ps = pstats.Stats(profiler, stream=s).sort_stats('cumulative')
        ps.print_stats(20)  # Top 20 functions
        print("\nTop CPU-consuming functions:")
        print(s.getvalue())

    if args.memprofile:
        snapshot = tracemalloc.take_snapshot()
        with open(args.memprofile, 'w') as f:
            f.write(f"Memory Profile\n")
            f.write(f"=" * 80 + "\n\n")
            top_stats = snapshot.statistics('lineno')
            f.write("Top 20 memory allocations:\n")
            for stat in top_stats[:20]:
                f.write(f"{stat}\n")
        tracemalloc.stop()
        print(f"Memory profile saved to {args.memprofile}")


if __name__ == "__main__":
    main()